MSR - Add option to change final boss

dyceron opened the pull request:
The final boss can now be changed to be Arachnus, Diggernaut, or Queen instead of Ridley. Logic has been updated to reflect this change.

~~**CURRENT ISSUE**: All seeds fail to generate, including Ridley seeds, and I can't figure out why.~~

![image](https://github.com/user-attachments/assets/10925cb2-5378-4137-8427-0d7710d2d8da)
![image](https://github.com/user-attachments/assets/a0f50bf9-deef-43c0-b0b6-9897faee6f4a)

dyceron commented:
Updated the DNA Placement section to mention the new changes

![image](https://github.com/user-attachments/assets/3c0171f3-0369-48a1-9a7f-f373f15fdd94)

Miepee commented on randovania/games/samus_returns/logic_database/Area 2 Dam Exterior.txt:
```diff
@@ -178,23 +174,37 @@ Extra - asset_id: collision_camera_005
+      Trivial
+  > Top of Wall
+      Single-wall Wall Jump (Intermediate) or Climb Rooms Vertically (No High Jump)
+  > Event - Credits
+      After Area 2 (Dam Exterior) - Arachnus
```
I would rename this node to `Event - Credits on Arachnus Goal` or something just to make it clearer that its on the layer without needing to click on it.

Miepee commented on randovania/games/samus_returns/logic_database/Area 2 Dam Exterior.txt:
```diff
@@ -178,23 +174,37 @@ Extra - asset_id: collision_camera_005
+      Trivial
+  > Top of Wall
+      Single-wall Wall Jump (Intermediate) or Climb Rooms Vertically (No High Jump)
+  > Event - Credits
```
when exactly does the game go to credits when beating arachnus? immediately after arachnus is dead? asking cause this seems like it would cause a mismatch of requirements.

dyceron replied:
When you re-enter Dam Exterior. I guess I only mentioned it patcher side and not here.

Miepee replied:
Gothca.

Miepee commented on randovania/games/samus_returns/logic_database/Area 2 Dam Exterior.txt:
```diff
@@ -261,7 +273,11 @@ Extra - asset_id: collision_camera_007
-      Lay Any Bomb
+      All of the following:
+          Lay Any Bomb
+          Any of the following:
+              Disabled Final Boss Arachnus
+              Enabled Final Boss Arachnus and All DNA
```
Whats the reasoning for this? I'm not following here.

dyceron replied:
To account for entering the room backwards from the tunnel in DLR

Miepee replied:
Feel like it might be cleaner to have those requirements in the arachnus arena itself but this is good enough for now.

Miepee commented on randovania/games/samus_returns/logic_database/Area 2 Dam Exterior.txt:
```diff
@@ -178,23 +174,37 @@ Extra - asset_id: collision_camera_005
       Trivial
 
-> Start Point (Bottom); Heals? False; Spawn Point
+> Next to Arachnus Arena; Heals? False; Spawn Point
```
noticed this while reviewing a6: any reason this isnt a start point?

dyceron replied:
Probably an oversight

dyceron replied:
Oh, wait I see what you mean. Honestly, I want to rename all the nodes named `Start Point` to be more descriptive, because it can be confusing for the rooms that have multple starting locations, since the name that shows up is `Start Point`

Miepee commented on randovania/games/samus_returns/logic_database/Area 6.txt:
```diff
@@ -765,6 +769,14 @@ Extra - asset_id: collision_camera_043
               Spider Ball
               # Lay a bomb near a weak point when its on the right or top only
               Mid-Air Morph (Advanced) or Infinite Bomb Jump (Advanced)
+  > Event - Credits
```
see Arachnus comment.

Miepee commented on randovania/games/samus_returns/logic_database/Area 6.txt:
```diff
@@ -813,12 +825,14 @@ Extra - asset_id: collision_camera_045
-              # Defeat the Gulluggs
-              Combat (Advanced) or Defeat Weak Enemies
+      All of the following:
+          Disabled Final Boss Diggernaut
```
why can you only collect the item if diggernaut is not the final boss?

dyceron replied:
The game ends after beating Diggernaut, so you never get to collect the item.

Miepee replied:
hmm, but theoretically if you would start behind diggernaut then you would be able to collect the item and beat diggernaut, no?

dyceron replied:
It's a bit more complicated then that. The item is "accessible" pre-fight, but requires OoB stuff. I probably should disable early access period.

Miepee replied:
In that theoretical case, would you be able to backdoor diggernaut?

dyceron replied:
The fight currently only starts if you enter from the right door. If you aren't OoB, you can't backdoor because of the wall in the arena.

Miepee commented on randovania/games/samus_returns/logic_database/Area 8.txt:
```diff
@@ -1187,9 +1187,15 @@ Extra - asset_id: collision_camera_020
-  > Dock to Hatchling Room
-      After Area 8 - Queen Metroid or Enabled Remove Queen Wall
+      All of the following:
+          After Area 8 - Larva Metroid 1 and After Area 8 - Larva Metroid 10 and After Area 8 - Larva Metroid 2 and After Area 8 - Larva Metroid 3 and After Area 8 - Larva Metroid 4 and After Area 8 - Larva Metroid 5 and After Area 8 - Larva Metroid 6 and After Area 8 - Larva Metroid 7 and After Area 8 - Larva Metroid 8 and After Area 8 - Larva Metroid 9
```
template?

dyceron replied:
I can see why you'd use one, but it's only used one time, so eh.

Miepee replied:
makes things easier if you need to move this again.

Miepee commented on randovania/games/samus_returns/logic_database/Surface West.txt:
```diff
@@ -197,7 +209,11 @@ Extra - asset_id: collision_camera_017
   > Dock to Landing Site
-      Baby Metroid and All DNA
+      All of the following:
+          Baby Metroid and Before Surface (West) - Proteus Ridley
+          Any of the following:
+              Disabled Final Boss Proteus Ridley
```
```suggestion
          Baby Metroid
          Any of the following:
              Disabled Final Boss Proteus Ridley and Before Surface (West) - Proteus Ridley
```

Miepee reviewed (commented):
I am fairly certain that you need to move spring ball and diggernauts PB also onto its own layer. The way it is right now, means that randovania will still try to place items on there, even if the requirement to reach them is Impossible. Which means that with an arachnus goal, you could see something like 

```
Space Jump
Area 2 Exterior - Arachnus Arena
```
in the credits.

---

I am also not a fan of making some of the users settings void without good clarification. Should probably have some GUI adjustments to it.

dyceron commented:
> I am fairly certain that you need to move spring ball and diggernauts PB also onto its own layer. The way it is right now, means that randovania will still try to place items on there, even if the requirement to reach them is Impossible. Which means that with an arachnus goal, you could see something like
> 
> ```
> Space Jump
> Area 2 Exterior - Arachnus Arena
> ```
> 
> in the credits.
> 

This would apply to all bosses then, no? Even then, is that even true? I feel like this is fine because they are misc resources and not just db paths.

> I am also not a fan of making some of the users settings void without good clarification. Should probably have some GUI adjustments to it.

I agree on adding more info to the GUI about the changes, since they aren't revealed to the player at all. Really can't do much about the preset invalidation in some parts so that would suffice.

Miepee commented on randovania/games/samus_returns/generator/bootstrap.py:
```diff
@@ -68,6 +78,17 @@ def _get_enabled_misc_resources(
+        # If Queen is the final boss, remove the wall next to the arena
+        elif configuration.final_boss == FinalBossConfiguration.QUEEN:
+            enabled_resources.add("FinalBossQueen")
+            enabled_resources.add("ReverseArea8")
```
But is the *change* mentioned anywhere?

dyceron replied:
I don't understand. If you're asking about GUI mentioning the wall removal if the Queen is the final boss, then no. Otherwise, it's already a thing.

Miepee replied:
was asking about the former. good to know.

Miepee commented on randovania/games/samus_returns/layout/msr_configuration.py:
```diff
@@ -44,11 +57,26 @@ class MSRConfiguration(BaseConfiguration):
     hints: HintConfiguration
     constant_heat_damage: int | None = dataclasses.field(metadata={"min": 0, "max": 1000, "precision": 1})
     constant_lava_damage: int | None = dataclasses.field(metadata={"min": 0, "max": 1000, "precision": 1})
+    final_boss: FinalBossConfiguration  # TODO: Add support to use random option
```
Ah, mustve misread then.

Miepee commented:
> This would apply to all bosses then, no? Even then, is that even true? I feel like this is fine because they are misc resources and not regular requirement paths.

Oh yeah it would apply to all bosses.
Don't forget that misc resources dont actually exist during generation; they're just replaced with Trivial or Impossible. The generator *will* place pickups on *every* pickupnode that exists, regardless of whether it's reachable or not. Having unreachable boss items on an existing layer thus means that the generator can place items on there. You can double check this by generating an arachnus seed and opening hte spoiler log and see whether arachnus spring ball appears.

dyceron commented:
Addressed the review. 

Updated the GUI on the `Optional Patches` tab.
![image](https://github.com/user-attachments/assets/6b274a39-985b-4e09-99ee-db29c9b2bde0)

Added the other layers for the final bosses if they aren't the final boss to ensure the pickups are relevant. Do I even need the misc resources anymore?

Miepee commented on randovania/games/samus_returns/generator/bootstrap.py:
```diff
@@ -37,14 +38,9 @@ def is_dna_node(node: PickupNode, config: BaseConfiguration) -> bool:
-            _boss_indices.pop(4)
+        for boss, index in _boss_mapping:
+            if config.final_boss.value == boss:
+                _boss_indices.pop(index)  # type: ignore
```
why is this a type ignore

dyceron replied:
Myoy was complaining

Miepee replied:
then fix it, and dont just ignore it.

dyceron replied:
I don't know what it wants 

```[mypy] reported by reviewdog 🐶
Argument 1 to "pop" of "list" has incompatible type "object"; expected "SupportsIndex"  [arg-type]
```

Miepee replied:
```py
_boss_mapping = [{"Arachnus": 0}, {"Diggernaut": 2}, {"Queen": 3}, {"Ridley": 4}]
```

Miepee replied:
and then change the for loop to be _boss_mapping.items()

dyceron replied:
Ugh how'd I not see that

dyceron replied:
I considered having the boss list only have the chase index, then appending all the valid indices depending on the final boss.

Miepee commented on randovania/games/samus_returns/gui/ui_files/preset_msr_patches.ui:
```diff
@@ -117,7 +117,7 @@
              </property>
              <property name="text">
-              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Removes the Grapple Blocks that are near the exit elevators of certain areas. These include:&lt;/p&gt;&lt;ul style=&quot;margin-top: 0px; margin-bottom: 0px; margin-left: 0px; margin-right: 0px; -qt-list-indent: 1;&quot;&gt;&lt;li style=&quot; margin-top:12px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent ...
+              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Removes the Grapple Blocks that are near the exit elevators of certain areas. These include:&lt;/p&gt;&lt;ul style=&quot;margin-top: 0px; margin-bottom: 0px; margin-left: 0px; margin-right: 0px; -qt-list-indent: 1;&quot;&gt;&lt;li style=&quot; margin-top:12px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent ...
```
change this to `...will always be removed even if this option is unchecked`.

Miepee commented on randovania/games/samus_returns/gui/ui_files/preset_msr_patches.ui:
```diff
@@ -177,7 +177,7 @@
              </property>
              <property name="text">
-              <string>Removes the wall between the Queen Arena and Hatchling Room, allowing for earlier access into Area 8 by entering the area backwards.</string>
+              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Removes the wall between the Queen Arena and Hatchling Room, allowing for earlier access into Area 8 by entering the area backwards.&lt;/p&gt;&lt;p&gt;If the Metroid Queen is the final boss, the wall will always be removed even if this option is disabled.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
same as with diggernaut

Miepee commented:
Additionally, you should also disable (not uncheck) the relevant checkboxes depending on the current final boss.

dyceron commented:
> Additionally, you should also disable (not uncheck) the relevant checkboxes depending on the current final boss.

It'd just be the one, but that's a good idea

Miepee commented:
Why one? Itd be diggernaut and queen, no?

dyceron commented:
Diggernaut is lumped in the Grapple Block removal patch

Miepee commented:
Ah. Hm.

dyceron commented:
Address the rest of the reviews.

If Queen is the final boss, the `Reverse Area 8` option is disabled.

![image](https://github.com/user-attachments/assets/de5eaec3-1ab2-4569-960f-978206efd595)
![image](https://github.com/user-attachments/assets/7c40afe2-eac3-47c5-8b5f-b51ee08bb2f1)

Also slightly updated the text for Diggernaut and Queen patches.

Miepee commented:
Can the dropdown info also mention that queen will remove the back wall / diggernaut wil remove the grapple block?

dyceron commented:
![image](https://github.com/user-attachments/assets/f91945a8-d2ea-4132-a9c0-5a6ec69a19a4)

![image](https://github.com/user-attachments/assets/f52b5692-f7f6-4263-80d8-acb88dd77a8f)

dyceron commented:
I'm unsure if that's a good spot since that part of the UI is used for mentioning how to access the bosses, and not other things related to the option.

Miepee commented:
The place feels fine for me, as I see it more as general info.But I would probably switch up the two lines for the queen (mention how to access first, then the wall)

dyceron commented:
I want to say that's everything covered?

Miepee commented:
I'll give it a rereview tomorrow. Since Thanatos chimed in too, I'll wait for him to rereview too.

dyceron commented on test/generator/test_generator_reach.py:
```diff
@@ -125,6 +125,8 @@ def _compare_actions(
     # Unknown reason why
     RandovaniaGame.CAVE_STORY: {30, 31, 41, 45},
+    # Ridley pickup only exists if Ridley is not the final boss
+    RandovaniaGame.METROID_SAMUS_RETURNS: {211},
```
Ah yeah this is a leftover change now
